password/src/Data/Password/Validate.hs

```diff
@@ -0,0 +1,276 @@
+   [ mconcat
+      [ validatePasswordPolicy passwordPolicy
+      , validateCharSetPredicate passwordPolicy charSetPredicate
+      ]
+   , catMaybes
+      [ isTooShort
+      , isTooLong
+      , isUsingValidCharacters
+      , hasRequiredChar uppercaseChars Uppercase
+      , hasRequiredChar lowercaseChars Lowercase
+      , hasRequiredChar specialChars Special
+      , hasRequiredChar digitChars Digit
+      ]
+   ]
+  where
+    -- Return first non-empty list
+    firstNonEmpty :: [[a]] -> [a]
+    firstNonEmpty [] = []
+    firstNonEmpty (x:xs)
+      | null x = firstNonEmpty xs
+      | otherwise = x
+    isTooLong :: Maybe InvalidReason
+    isTooLong =
+      if T.length password <= maximumLength
+        then Nothing
+        else Just $ PasswordTooLong maximumLength (T.length password)
+    isTooShort :: Maybe InvalidReason
+    isTooShort =
+      if T.length password >= minimumLength
+          then Nothing
+          else Just $ PasswordTooShort minimumLength (T.length password)
+    isUsingValidCharacters :: Maybe InvalidReason
+    isUsingValidCharacters =
+        let filteredText = T.filter (\c -> not $ (getCharSetPredicate charSetPredicate) c) password
+        in if T.null filteredText
+          then Nothing
+          else Just $ InvalidCharacters filteredText
+    hasRequiredChar :: Maybe Int -> CharacterCategory -> Maybe InvalidReason
+    hasRequiredChar Nothing _ = Nothing
+    hasRequiredChar (Just requiredCharNum) characterCategory =
+      let predicate = categoryToPredicate characterCategory
+          actualRequiredCharNum = T.length $ T.filter predicate password
+       in if actualRequiredCharNum >= requiredCharNum
+          then Nothing
+          else Just $ NotEnoughReqChars characterCategory requiredCharNum actualRequiredCharNum
+
+-- | Validate 'CharSetPredicate' that it returns 'True' on at least one of the characters
+-- that is required
+--
+-- For instance, if 'PasswordPolicy' states that the password requires at least
+-- one uppercase letter, then 'CharSetPredicate' should return True on at least
+-- one uppercase letter.
+validateCharSetPredicate :: PasswordPolicy -> CharSetPredicate -> [InvalidReason]
+validateCharSetPredicate PasswordPolicy{..} (CharSetPredicate predicate) =
+  let charSets = accumulateCharSet
+        [ (uppercaseChars, Uppercase)
+        , (lowercaseChars, Lowercase)
+        , (specialChars, Special)
+        , (digitChars, Digit)
+        ]
+  in catMaybes $ map checkPredicate charSets
+  where
+    checkPredicate :: (Int, CharacterCategory, String) -> Maybe InvalidReason
+    checkPredicate (num, category, sets) =
+      if any predicate sets
+        then Nothing
+        else Just $ InvalidCharSetPredicate category num
+    accumulateCharSet :: [(Maybe Int, CharacterCategory)] -> [(Int, CharacterCategory, String)]
+    accumulateCharSet = map (\(num, c) -> (num, c, categoryToString c))
+      . filter ((> 0) . fst)
+      . map (\(mnum, c) -> (fromMaybe 0 mnum, c))
```

Comment:
This might be more readable:
```hs
accumulateCharSet xs =
  [ (num, c, categoryToString c)
  | (mNum, c) <- xs
  , num <- fromMaybe 0 mNum
  , num > 0
  ]
```